src/Lykke.AzureStorage/Tables/Decorators/EncryptedTableStorageDecorator.cs

```diff
@@ -0,0 +1,461 @@
+using Microsoft.WindowsAzure.Storage.Table;
+
+namespace AzureStorage.Tables.Decorators
+{
+    /// <summary>
+    /// Wrapper on INoSQLTableStorage, uses symmetric encryption.
+    /// Does not support queries (TableQuery<T>) and batches(TableBatchOperation).
+    /// </summary>
+    public class EncryptedTableStorageDecorator<T> : INoSQLTableStorage<T> where T : ITableEntity, new()
+    {
+        private readonly INoSQLTableStorage<T> _storage;
+        private readonly ICryptographicSerializer _serializer;
+        private readonly bool _forceEncrypt;
+        private readonly List<PropertyInfo> _encryptedProperties;
+
+        /// <summary>
+        /// Ctor.
+        /// </summary>
+        /// <param name="storage"></param>
+        /// <param name="serializer"></param>
+        /// <param name="forceEncrypt">If true, checks if existed entires are encrypted, and encrypt them if not.</param>
+        public EncryptedTableStorageDecorator(
+            [NotNull] INoSQLTableStorage<T> storage,
+            [NotNull] ICryptographicSerializer serializer,
+            bool forceEncrypt = true)
+        {
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+            _serializer = serializer ?? throw new ArgumentNullException(nameof(serializer));
+            _forceEncrypt = forceEncrypt;
+
+            var encryptAttribute = typeof(EncryptAttribute);
+            _encryptedProperties = typeof(T).GetProperties().Where(x => Attribute.IsDefined(x, encryptAttribute)).ToList();
+            if (_encryptedProperties.Count == 0)
+                throw new ArgumentException("No properties marked as encrypted.");
+            if (_encryptedProperties.Any(x => x.PropertyType != typeof(string)))
+                throw new ArgumentException($"Only {typeof(string).FullName} type properties can be marked as encrypted.");
+            if (_encryptedProperties.Any(x => x.SetMethod == null || x.GetMethod == null))
+                throw new ArgumentException("Only properties with both Get and Set method allowed. Please check this: " + string.Join(", ",
+                                                _encryptedProperties.Where(x => x.SetMethod == null || x.GetMethod == null).Select(x => x.Name)));
+        }
+
+        public static INoSQLTableStorage<T> Create(INoSQLTableStorage<T> storage, ICryptographicSerializer algo)
+        {
+            return new EncryptedTableStorageDecorator<T>(storage, algo);
+        }
+
+        private T Decrypt(T entity)
+        {
+            if (entity == null)
+            {
+                return default(T);
+            }
+
+            var hasNonEncryptedProperties = false;
+            foreach (var property in _encryptedProperties)
+            {
+                var encrypted = property.GetValue(entity) as string;
+                if (string.IsNullOrEmpty(encrypted))
+                    continue;
+
+                if (_forceEncrypt && !_serializer.IsEncrypted(encrypted))
```

Comment:
Looks like _forceEncrypt chould be checked only at the line 83

Reply:
In fact, this check is needed only here, and not needed at the line 83. Updated.